Use system glslang

psi29a opened the pull request:
## Description

Allow support of using system's glslang if available, otherwise fallback to original git based gslang method.

Fixes https://github.com/vsgopenmw-dev/vsgopenmw/issues/9

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?



## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

robertosfield committed:
Initial experiment with Tracy style instrumentation

robertosfield committed:
Windows build fix

psi29a committed:
make use of system glslang if found

robertosfield committed:
Created a vsg::Instrumentation base class

psi29a committed:
bump plugin versions

psi29a committed:
bump cmake to 3.24 to get all the nice cmake magic

psi29a committed:
make use of Vulkan::glslang

robertosfield committed:
Made vsg::Instrumentation a pure virtual base class and added commandBuffer parameter.
Added reference value to the Instrumentation::enter(..) & leave(..) methods

robertosfield committed:
Added assignment of colours

robertosfield committed:
Implemented a VulkanAnnotation version of vsg::Instrumentation to enable population of Vulkan API layer/RenderDoc in a more useful way.

robertosfield committed:
Added instrumentation hooks to Viewer, RecordAndSubmitTask & CommandGraph

robertosfield committed:
Merge pull request #1052 from vsg-dev/Intrumentation

Instrumentation

robertosfield committed:
Moved include to address Windows build error

robertosfield committed:
Ran clang-format

timoore committed:
Spelling fix

timoore committed:
Add rump API for initializing Tracy GPU profiling

timoore committed:
Merge branch 'Intrumentation' into Instrumentation

robertosfield committed:
Merge pull request #1053 from timoore/Instrumentation

Instrumentation

robertosfield committed:
Added instrumentation support for TransferTask

robertosfield committed:
Moved the Instrumentation::init() call to the Viewer::compile().

AnyOldName3 commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.7)
+cmake_minimum_required(VERSION 3.24)
```
Have you checked nothing would be affected by any of the policy changes between 3.7 and 3.24. Changing `cmake_minimum_required` changes all policies introduced in the version range from their `OLD` behaviour to their `NEW` behaviour.

psi29a replied:
There have been no changes (oddly enough), here is the complete output:

```
/Applications/CLion.app/Contents/bin/cmake/mac/aarch64/bin/cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_MAKE_PROGRAM=/Applications/CLion.app/Contents/bin/ninja/mac/aarch64/ninja -G Ninja -S /Users/psi29a/Workspace/private/VulkanSceneGraph -B /Users/psi29a/Workspace/private/VulkanSceneGraph/cmake-build-relwithdebinfo
-- The CXX compiler identification is AppleClang 15.0.0.15000040
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Reading 'vsg_...' macros from /Users/psi29a/Workspace/private/VulkanSceneGraph/cmake/vsgMacros.cmake - look there for documentation
-- Found Vulkan: /opt/homebrew/lib/libvulkan.dylib (found suitable version "1.3.268", minimum required is "1.1.70.0") found components: glslang SPIRV-Tools glslc glslangValidator 
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- The following REQUIRED packages have been found:

 * Vulkan (required version >= 1.1.70.0)
 * Threads

-- Configuring done (4.8s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/psi29a/Workspace/private/VulkanSceneGraph/cmake-build-relwithdebinfo

[Finished]
```

AnyOldName3 replied:
This changes the behaviour of 72 CMake policies. It's unlikely that none of them would have affected anything. The differences might only end up breaking setups you've not tried, or might only result in runtime changes. There're probably a few changes that mean something that's being handled explicitly can be simplified as CMake will do it implicitly. You really need to look through the policy list https://cmake.org/cmake/help/latest/manual/cmake-policies.7.html#manual:cmake-policies(7) and check each one rather than just try building and assume everything is fine if it looks like it worked.

psi29a replied:
> The differences might only end up breaking setups you've not tried

This is a sure fire way to prevent any commit from being merged. I hope you understand that.

Which is why we try to make sure the CI is as robust as possible, within reason. It is impossible to test every place that VSG or any library would be used on.

Thanks for the link, I'll review.

Building is not enough, I've validated on 3 platforms except Windows. So I'll need someone's help making sure that it actually runs there.

AnyOldName3 replied:
> This is a sure fire way to prevent any commit from being merged. I hope you understand that.

Yes. Changing what's passed to `cmake_minimum_required` is a big deal, and needs an abundance of caution. It's not as minor as doing something like changing the C++ language standard, and that's already a pretty big deal.

robertosfield committed:
Added support for signally the start and end of command buffer recording

robertosfield committed:
Added instrumentation of Bin

robertosfield committed:
Added VertexIndexDraw, VertexDraw and Geometry to list of RecordTraversal::apply() to enable more useful instrumentation

robertosfield committed:
Added Viewer::assignInstrumentation(..) method to help with setting up the viewer and associated objects with Instrumentation

robertosfield committed:
Added Instrumentation::enterFrame() and leaveFrame().

robertosfield committed:
Build fix for Windows

robertosfield committed:
Replaced ref_ptr<> with reference

robertosfield committed:
Added VSG_MAX_INSTRUMENTATION_LEVEL cmake controlled #define for controlling what level of instrumentation is compiled into the VSG

robertosfield committed:
Added instrumentation laevel to marco usage

robertosfield committed:
Renamed instrumentation macros to better reflect that function

robertosfield committed:
Added CPU/GPU_INSTRUMENTATION_L1, _L2, _L3 marcros

robertosfield committed:
refined class naming and moved GpuAnnotation into it's own header/source

robertosfield committed:
Added header online TracyInstrumentation header to eanble use of Tracy witout the the VSG itself being compiled against Tracy - only client applications that want to use need to.

robertosfield committed:
Removed unneccessary include

robertosfield committed:
Added an removal of TracyInstrumentation.h from the build_all_h to prevent the Tracy dependency spilling over into applications that don't use it.

robertosfield committed:
Cleanup and ran clang-format

robertosfield committed:
Removed debug message

robertosfield committed:
Implemented support for compiling with/out TRACY_ON_DEMAND enabled

robertosfield committed:
Added CommandBuffer instrumentation

robertosfield committed:
Added instrumentation to CompileManager/CompileTraversal/Context

robertosfield committed:
Added vsg::Instrumentation support to DatabasePager

robertosfield committed:
Added support for duplicating Instrumentation when required to enable multi-threaded instrumentation via multiple Instrumentation objects

robertosfield committed:
Ran clang-format

robertosfield committed:
Fixed overrriding of locally assign Instrumentation when is not assigned

robertosfield committed:
Added instumentation to vsg::Options

robertosfield committed:
Added Instrumentation support to IO operations

robertosfield committed:
Added timestamp calaibrartion extensions and prelimnary support in TracyInstrumentation.

robertosfield committed:
Removed debug messages

robertosfield committed:
Added support for TRACY_ENABLE being defined/or not.

robertosfield committed:
Replaced Device* with ref_ptr<> to prevent Instrumentation not keepig the vk/vsg::Device around long enough to clean up the VkXtx related Vulkan objects

robertosfield committed:
Removed debug messages

robertosfield committed:
Added Instrumentation::setThreadName(const std::string&) support.

robertosfield committed:
Added name and colour entries for main operations

robertosfield committed:
Fixed name

robertosfield committed:
Changed to using more efficient VertexIndexDraw

robertosfield committed:
Removed hardwired enabling of GpuAnnotation.

robertosfield committed:
Introduce a uint_color helper struct to make it easier to manage colours that work with Tracy and with the VSG/Vulkan.

robertosfield committed:
Ran clang-format

robertosfield committed:
Added GpuAnnotation::labelType to hint which member of SourceLocation to use as the VK_debug_utils label.

robertosfield committed:
Moved the #define COLOR to sit before the rest of the #define

robertosfield committed:
Added instrumentation of TileDatabase to RecordTraversal.

robertosfield committed:
Refined colours

robertosfield committed:
Added more colour differentiation

robertosfield committed:
Added instrumentation to ViewDependentState

robertosfield committed:
Resutrctured the TracyVkCollect call so it's done once per frame as the on the first command buffer record of a frame.

robertosfield committed:
Updated SHaderSet after fix to standard.vert's binding of the displacement map

robertosfield committed:
Merge branch 'master' into Instrumentation

jamierobertson1 committed:
Update UpdateOperations.cpp

Prevents a crash in visual studio when in Debug mode when removing a sole update operation from the list. (debug assertion: "cannot increment end list iterator").

jamierobertson1 committed:
Update UpdateOperations.cpp

robertosfield committed:
Merge pull request #1059 from jamierobertson1/master

Update UpdateOperations::Remove to prevent Visual Studio crash in debug mode

robertosfield committed:
Merge branch 'master' into Instrumentation

robertosfield committed:
Changed Window::clearColor() to use a vsg::vec4.
Added sRGB_to_linear(vec4) and linear_to_sRGB(vec4) convinience functions.
Added vec4::operator VkClearColorValue() to help wuth vec4/ClearColor compatibility.

robertosfield committed:
Added support for assigning VkClearColorValue to vsg::vec4

robertosfield committed:
Fixed exponents

robertosfield committed:
Added sRGB converted colour to debug output

robertosfield committed:
Changed the handling of alpha to remain linear in all cases

robertosfield committed:
Merge branch 'master' into Instrumentation

robertosfield committed:
Ran clang-format

robertosfield committed:
Changed rc to 0 for 1.1.0 release

robertosfield committed:
Updated version

robertosfield committed:
Added const Object* to Instrumentation::enter()/leave() methods

robertosfield committed:
Added explict

robertosfield committed:
Merge branch 'master' into Instrumentation

robertosfield committed:
Improved doxygen comments

robertosfield committed:
Merge branch 'Instrumentation' of https://github.com/vsg-dev/VulkanSceneGraph into Instrumentation

robertosfield committed:
Added enabledExtensions to vsg::Device and cleaned up placement of extensions functions

robertosfield committed:
Changed to using the new Device::supportsDeviceExtensions(..)

robertosfield committed:
Added vsg::InstrumentationNode

robertosfield committed:
Added InstrumentationNode support and warning fixes

robertosfield committed:
Fixed copyright date

robertosfield committed:
Improved comment

robertosfield committed:
Merge pull request #1062 from vsg-dev/Instrumentation

Added vsg::Instrumentation scheme to enable extensible instrumentation.

robertosfield committed:
Fixed typo

john-tecplot committed:
FIX: handle change aspect change when render area is not full window.

When the camera viewport is not the full window, the aspect ratio of the perspective
projection matrix needs to be adjusted in the same was as the Ellipsoid case.

robertosfield committed:
Merge pull request #1066 from theodoregoetz/fix-perspective-resize-subarea

FIX: handle change aspect change when render area is not full window.

robertosfield committed:
Added check for valid array pointer

robertosfield committed:
Merge branch 'master' of https://github.com/vsg-dev/VulkanSceneGraph

robertosfield committed:
Added explicit

john-tecplot committed:
add deps to allow inclusion of io/Path.h directly

robertosfield committed:
Added definition of type_name<const Array<T>>

robertosfield committed:
Merge pull request #1073 from theodoregoetz/make-path-includable

add deps to allow inclusion of io/Path.h directly

jamierobertson1 committed:
Fix keyboard modifier typos in Update Win32_Window.h

Typos in vsgWin32::KeyboardMap::getKeySymbol prevented VK_RSHIFT and VK_RMENU key's (Right shift & Right Alt) being mapped to MODKEY_Shift and MODKEY_Alt keys.

jamierobertson1 committed:
Merge pull request #1 from jamierobertson1/jamierobertson1-patch-1

Fix keyboard modifier typos in Update Win32_Window.h

robertosfield committed:
Merge pull request #1076 from jamierobertson1/master

Fix keyboard modifier typos in Update Win32_Window.h

robertosfield committed:
Created initial data structures to store animation related data loaded from vsgXchange::assimp

robertosfield committed:
Bumped version to 1.1.2 to mark the beginning of animation classes

robertosfield committed:
Added export

robertosfield committed:
Improved compatibility with older vulkan header versions

robertosfield committed:
Merge branch 'master' into Animation

robertosfield committed:
Implemented basic transform and morph animation data structures

robertosfield committed:
Removed no longer used class

robertosfield committed:
Fixed defintions to work with the appropriate Vulkan headers

robertosfield committed:
Fixed defintions to work with the appropriate Vulkan headers

robertosfield committed:
Merge branch 'master' into Animation

robertosfield committed:
Added iniital cut of AnimationTransform node and renamed the keyframe data classes.

robertosfield committed:
Added vsg::RiggedTransform for transforms that have nu mesh benath them so don't directly affect rendering, instead a used to control skinning bone matrices

robertosfield committed:
Added link to rocky.

robertosfield committed:
Added support for the new Animation objects to Visitor/ConstVisitor/RecordTraversal

robertosfield committed:
Added simulationTime to FrameStamp

robertosfield committed:
Added AnimationGroup/Animation::update(double simulationTime) methds

robertosfield committed:
Added update methods

robertosfield committed:
Implemented crude tracking of position

robertosfield committed:
Implemented animation of transforms

robertosfield committed:
Added AnimationManager

robertosfield committed:
Added AnimationManager to Viewer by default

robertosfield committed:
Added include to fix Windows build

robertosfield committed:
Fixed dates

robertosfield committed:
Ran clang0-format

robertosfield committed:
Added spport for mat4Value and dmat4Value to Visitors

robertosfield committed:
Added mat3/4::set methods

robertosfield committed:
Added TransformSampler

robertosfield committed:
Created a vsg::AnimationSampler baser class, with TransformSampler and MorphSampler subclasses.
Moved various animation classes out into their own dedicated header/source files.

robertosfield committed:
Added missing settinf of defines

robertosfield committed:
Added JointSampler

robertosfield committed:
Added JointSampler and updated built-in ShaderSets

robertosfield committed:
Added vsg::SetPrimitiveTopology command

robertosfield committed:
Added vsg::SetPrimitiveTopology command

robertosfield committed:
Added SetPrimitiveTopology command

robertosfield committed:
Added SetPrimitiveTopology command

robertosfield committed:
Merge branch 'master' into Animation

robertosfield committed:
Added extenion support for VK_EXT_extended_dynamic_state

robertosfield committed:
Fixed types

robertosfield committed:
Ran clang-format

robertosfield committed:
Merge branch 'master' into Animation

robertosfield committed:
Ran clang-format

robertosfield committed:
Removed AnimationTransform.
Renamed RiggedTransform to Joint and moved it to it's own header/source file.

robertosfield committed:
Refined naming and assignment of matrices

robertosfield committed:
Added build_simple_ShaderSets build target to enable easier testing/debugging of built-in shaders

robertosfield committed:
Added traverse implementations and updated ShaderSets

robertosfield committed:
Fixed transform ordering

robertosfield committed:
Added joint index to vsg::Joint

robertosfield committed:
Added missing override

robertosfield committed:
Removed unused variable

robertosfield committed:
Ran clang-format

robertosfield committed:
Added indentation helper class from vsgXchange

robertosfield committed:
Fixed warning

robertosfield committed:
Added Instrumentation to AnimationManager

robertosfield committed:
Generalized the type of node that can be a subgraph of a JointSampler

robertosfield committed:
Added FindAnimations

robertosfield committed:
Added FindAnimations

robertosfield committed:
Renamed AnimationManager::start(..) to play(..), and end(..) to stop(..).
Added automatic removing of completed animations from the AnimationManager.

robertosfield committed:
Addded AnimationManager::stop() method and doxygen commants

robertosfield committed:
Changed the default mode to repeat

robertosfield committed:
Changed the serialization order of members to make .vsgt more readable

robertosfield committed:
Added Visitor suppot for AnimationSamplers

robertosfield committed:
Added traversal of stateCommands to StateGroup.
Removed the now rendent local implementations of traversing stateCommands

robertosfield committed:
Implemented Object::clone(CopyOp&) for classes relevant to clone animation scene graph

robertosfield committed:
Added missing cloning of members

robertosfield committed:
Changes struct to class

robertosfield committed:
Experiment with providing a copy constructor.

robertosfield committed:
Commanded out copy constructor

robertosfield committed:
Added Animation copy constructor

robertosfield committed:
Implemented a better deafult Object::clone(copyop) implemenntation

robertosfield committed:
Ran clang-format

robertosfield committed:
Restructured CopyOp so that it is passed by const reference with a default constructor.

robertosfield committed:
Merge branch 'AnimationCopyConstructor' of https://github.com/vsg-dev/VulkanSceneGraph into AnimationCopyConstructor

robertosfield committed:
Ran clang-format

robertosfield committed:
Removede CopyOp.h header

robertosfield committed:
Implemented copy constructors

robertosfield committed:
Ran clang-format

robertosfield committed:
Simplified code

robertosfield committed:
Added copy constructor

robertosfield committed:
Added copy constructor

robertosfield committed:
Added copy constructor

robertosfield committed:
Refined BufferInfo::clone() implementation

robertosfield committed:
cppecheck fixes

robertosfield committed:
Fixed cpp-check supporession

robertosfield committed:
Added map accessors to Duplicate container

robertosfield committed:
Added ImageView, Image and ImageInfo support to Visitor and ConstVisitor

robertosfield committed:
Implemented FindDynamicObjects and PropogateDynamicObjects visitors.

robertosfield committed:
Added use of FindDynamicObjects, PropogateDynamicObjects and CopOp to enable automatic cloning of parts of scene graph that are dynamic when loading objects from SharedObjects.

robertosfield committed:
Added FindDynamicObjects and PropagateDynamicObjects to Options to enable users to specify their own mechanism for determining what parts of the scene graph are dynamic and need to be cloned during sharing of loaded models that contain animations etc.

robertosfield committed:
Ran clang-format

robertosfield committed:
Seperated FindDynamicObjects and PropagateDynamicObjects out into their own header and source files

robertosfield committed:
Ran clang-format

robertosfield committed:
Added initializer

robertosfield committed:
Added include to fix Windows build

robertosfield committed:
Fixed initialization order

robertosfield committed:
Added nullptr check

robertosfield committed:
Improved support for SecondaryCommandGraph addressing a regression/crash in the process

robertosfield committed:
Merge branch 'master' into AnimationCopyConstructor

robertosfield committed:
Ran clang-format

robertosfield committed:
Merge branch 'master' into AnimationCopyConstructor

robertosfield committed:
Implemented copy constructors & compare methods

robertosfield committed:
Added copy constructors

robertosfield committed:
Ran clang-format

robertosfield committed:
Added copy constructors

robertosfield committed:
Implemented copy constructors and compare

robertosfield committed:
Added copy constructors

robertosfield committed:
Ran clang-format

robertosfield committed:
Added copy constructor and compare methods

robertosfield committed:
Ran clang-format

robertosfield committed:
Added copy constructors, clone and compare

robertosfield committed:
Implmeneted copy constructor, clone compare and read/write methods

robertosfield committed:
Ran clang-format

robertosfield committed:
Implemented compare method

robertosfield committed:
Fixed recursive call bug

robertosfield committed:
Standardized positioning of overriden methods

robertosfield committed:
Changed linear search to binary search

robertosfield committed:
Added Animation::start()/stop() and caching of the start time and max time.

robertosfield committed:
Ran clang-format

robertosfield committed:
Fixed sample method

robertosfield committed:
Ran clang-format

robertosfield committed:
Suppressed silly cppcheck messages

robertosfield committed:
Added suppression of inapprorpiate cppcheck report

robertosfield committed:
Removed dev comment

robertosfield committed:
Added call to stop() when update is called when time exceeds maxTime.

robertosfield committed:
Updated built-in shder sets

robertosfield committed:
Added doxygen comments

robertosfield committed:
Added handling of dynamic dataVariance.

Mikalai committed:
Add is_compatible to Visitor and ConstVisitor

robertosfield committed:
Merge pull request #1093 from Mikalai/1089

Add is_compatible to Visitor and ConstVisitor

robertosfield committed:
Merge branch 'master' into AnimationCopyConstructor

robertosfield committed:
Ran clang-format

robertosfield committed:
Merge branch 'AnimationCopyConstructor' of https://github.com/vsg-dev/VulkanSceneGraph into AnimationCopyConstructor

robertosfield committed:
Implmented support for negative speeds and changing speed during animation.

robertosfield committed:
Ran clang-format

robertosfield committed:
Merge pull request #1094 from vsg-dev/AnimationCopyConstructor

Animation and object cloning/copy constrcutors

lufriem committed:
Typo

robertosfield committed:
Merge pull request #1096 from LeeRiva/lrdev

Typo

robertosfield committed:
Moved the animation time value into public scope and added the inital start time to the Animation::start(..) method

robertosfield committed:
Fixed regression in interesections

robertosfield committed:
Added ability to pass the simulationTime to use

robertosfield committed:
Updated ShaderSets

robertosfield committed:
Fixed CopyOp's handling of Auxiliary data.

robertosfield committed:
Improved grammer
